### ISSUE TITLE
Panic with `hugeParamChecker` with generic struct

### ISSUE THREAD

**Issue by nknordeen** [go-critic/go-critic#1298]
Hi, I'm seeing this panic:

```
level=error msg="[runner] Panic: gocritic: package \"endpoints\" (isInitialPkg: true, needAnalyzeSource: true): go/types/sizes.go:82: assertion failed: goroutine 11372 [running]:
runtime/debug.Stack()
	runtime/debug/stack.go:24 +0x64
github.com/golangci/golangci-lint/pkg/golinters/goanalysis.(*action).analyzeSafe.func1()
	github.com/golangci/golangci-lint/pkg/golinters/goanalysis/runner_action.go:102 +0x104
panic({0x1012f84c0, 0x14002a3e9c0})
	runtime/panic.go:884 +0x204
go/types.assert(0xe8?)
	go/types/errors.go:27 +0x60
go/types.(*StdSizes).Alignof(0x140001af150, {0x1014c95e8?, 0x14004534d50})
	go/types/sizes.go:82 +0x1a8
go/types.(*StdSizes).Offsetsof(0x1014c9458?, {0x140039e9cd0, 0x1, 0x0?})
	go/types/sizes.go:123 +0xc8
go/types.(*StdSizes).Sizeof(0x140001af150, {0x1014c9458?, 0x14005ad17a0})
	go/types/sizes.go:176 +0x278
github.com/go-critic/go-critic/framework/linter.(*CheckerContext).SizeOf(...)
	github.com/go-critic/go-critic@v0.6.5/framework/linter/linter.go:325
github.com/go-critic/go-critic/checkers.(*hugeParamChecker).checkParams(0x14004a93578, {0x1400787df38, 0x1, 0x101485220?})
	github.com/go-critic/go-critic@v0.6.5/checkers/hugeParam_checker.go:56 +0x11c
github.com/go-critic/go-critic/checkers.(*hugeParamChecker).VisitFuncDecl(0x14004e68c08?, 0x14005610060)
	github.com/go-critic/go-critic@v0.6.5/checkers/hugeParam_checker.go:44 +0x38
github.com/go-critic/go-critic/checkers/internal/astwalk.(*funcDeclWalker).WalkFile(0x14004e6e0e0, 0x14007a05080)
	github.com/go-critic/go-critic@v0.6.5/checkers/internal/astwalk/func_decl_walker.go:21 +0xec
github.com/go-critic/go-critic/framework/linter.(*Checker).Check(...)
	github.com/go-critic/go-critic@v0.6.5/framework/linter/linter.go:130
github.com/golangci/golangci-lint/pkg/golinters.runGocriticOnFile(0x14004fae240, 0x1400074e5e6?, {0x1400b04d800, 0x68, 0x10136a700?})
	github.com/golangci/golangci-lint/pkg/golinters/gocritic.go:175 +0xa0
github.com/golangci/golangci-lint/pkg/golinters.runGocriticOnPackage(0x14004fae240, {0x1400b04d800, 0x68, 0x80}, {0x1400490c740, 0x7, 0x140026a9b98?})
	github.com/golangci/golangci-lint/pkg/golinters/gocritic.go:163 +0x15c
github.com/golangci/golangci-lint/pkg/golinters.(*goCriticWrapper).run(0x14001020a50, 0x14005fa8000)
	github.com/golangci/golangci-lint/pkg/golinters/gocritic.go:123 +0x1c8
github.com/golangci/golangci-lint/pkg/golinters.NewGoCritic.func1(0x1013622a0?)
	github.com/golangci/golangci-lint/pkg/golinters/gocritic.go:46 +0x38
github.com/golangci/golangci-lint/pkg/golinters/goanalysis.(*action).analyze(0x14002c222d0)
	github.com/golangci/golangci-lint/pkg/golinters/goanalysis/runner_action.go:188 +0x958
github.com/golangci/golangci-lint/pkg/golinters/goanalysis.(*action).analyzeSafe.func2()
	github.com/golangci/golangci-lint/pkg/golinters/goanalysis/runner_action.go:106 +0x20
github.com/golangci/golangci-lint/pkg/timeutils.(*Stopwatch).TrackStage(0x140021b0fa0, {0x100fd77d2, 0x8}, 0x140029d4730)
	github.com/golangci/golangci-lint/pkg/timeutils/stopwatch.go:111 +0x44
github.com/golangci/golangci-lint/pkg/golinters/goanalysis.(*action).analyzeSafe(0x100a33050?)
	github.com/golangci/golangci-lint/pkg/golinters/goanalysis/runner_action.go:105 +0x74
github.com/golangci/golangci-lint/pkg/golinters/goanalysis.(*loadingPackage).analyze.func2(0x14002c222d0)
	github.com/golangci/golangci-lint/pkg/golinters/goanalysis/runner_loadingpackage.go:80 +0xb0
created by github.com/golangci/golangci-lint/pkg/golinters/goanalysis.(*loadingPackage).analyze
	github.com/golangci/golangci-lint/pkg/golinters/goanalysis/runner_loadingpackage.go:75 +0x168
"
level=warning msg="[runner] Can't run linter goanalysis_metalinter: goanalysis_metalinter: gocritic: package \"endpoints\" (isInitialPkg: true, needAnalyzeSource: true): go/types/sizes.go:82: assertion failed"
level=error msg="Running error: 1 error occurred:
	* can't run linter goanalysis_metalinter: goanalysis_metalinter: gocritic: package \"endpoints\" (isInitialPkg: true, needAnalyzeSource: true): go/types/sizes.go:82: assertion failed

"
```

The code that caused it to fail was this:
```golang
type mockConfig[T any] struct {
	cfg T
}

func (m mockConfig[T]) RegisterCallback(_ dcs.Callback[T]) {}

func (m mockConfig[T]) StartPoll() {}

func (m mockConfig[T]) GetCurrentConfig() *T {
	return &m.cfg
}
```

The panic stopped happening when I updated the code to this:

```golang
type mockConfig[T any] struct {
	cfg T
}

func (m *mockConfig[T]) RegisterCallback(_ dcs.Callback[T]) {}

func (m *mockConfig[T]) StartPoll() {}

func (m *mockConfig[T]) GetCurrentConfig() *T {
	return &m.cfg
}
```

**Comment by nknordeen** [comment]
I'm not sure if it's related to this: https://github.com/go-critic/go-critic/issues/1297 or if they're different

**Comment by mwchambers** [comment]
Here's a smaller reproduction of the issue:

```go
type Maybe[V any] struct {
    value V
}
func (m Maybe[V]) Fn() {}
```

I think this is similar to what is addressed in #1237

Though for this case the type is not a typeparams.TypeParam. 

It is a `types.Named` with non-nil `TypeParamList`

So this might work:

```go
// Existing check:
if _, ok := typ.(*typeparams.TypeParam); ok {
    continue
}

// Also add this?
if named, ok := typ.(*types.Named); ok && named.TypeParams() != nil {
   continue
}
```

Though it might be best to do a bit more digging into how `go/types` would normally prevent `SizeOf` hitting the assertion:

https://cs.opensource.google/go/go/+/refs/tags/go1.20:src/go/types/sizes.go;l=181

**Comment by cristaloleg** [comment]
Should be fixed by #1309 

Next week we will release a new version of `go-critic` with this and other fixes. Thank everyone.
